File: kubernetes-informer/src/test/java/io/micronaut/kubernetes/client/informer/ConfigMapInformer.java
/*
 * Copyright 2021 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.kubernetes.client.informer;

import io.kubernetes.client.informer.ResourceEventHandler;
import io.kubernetes.client.openapi.models.V1ConfigMap;
import io.kubernetes.client.openapi.models.V1ConfigMapList;
import io.micronaut.context.annotation.Requires;
import jakarta.inject.Singleton;
import org.apache.commons.compress.utils.Lists;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

@Requires(property = "spec.name", value = "ConfigMapInformerSpec")
//tag::handler[]
@Singleton
@Informer(apiType = V1ConfigMap.class, apiListType = V1ConfigMapList.class) // <1>
public class ConfigMapInformer implements ResourceEventHandler<V1ConfigMap> { // <2>

    //end::handler[]
    public static final Logger LOG = LoggerFactory.getLogger(ConfigMapInformer.class);

    private final List<V1ConfigMap> added = Lists.newArrayList();
    private final List<V1ConfigMap> updated = Lists.newArrayList();
    private final List<V1ConfigMap> deleted = Lists.newArrayList();

    public List<V1ConfigMap> getAdded() {
        return added;
    }

    public List<V1ConfigMap> getUpdated() {
        return updated;
    }

    public List<V1ConfigMap> getDeleted() {
        return deleted;
    }

    //tag::handler[]
    @Override
    public void onAdd(V1ConfigMap obj) {
        //end::handler[]
        LOG.info("ADDED CONFIG MAP: {}", obj);
        added.add(obj);
        //tag::handler[]
    }

    @Override
    public void onUpdate(V1ConfigMap oldObj, V1ConfigMap newObj) {
        //end::handler[]
        LOG.info("UPDATED CONFIG MAP: {}", newObj);
        updated.add(newObj);
        //tag::handler[]
    }

    @Override
    public void onDelete(V1ConfigMap obj, boolean deletedFinalStateUnknown) {
        //end::handler[]
        LOG.info("DELETED CONFIG MAP: {}", obj);
        deleted.add(obj);
        //tag::handler[]
    }
}
//end::handler[]

File: test-suite-python/src/test/python/micronaut/docs/informer/ConfigMapInformer.py
import logging

from micronaut.context.annotation import Requires
from micronaut.kubernetes.client.informer import Informer

try:
    from io.kubernetes.client.informer import ResourceEventHandler
    from io.kubernetes.client.openapi.models import V1ConfigMap, V1ConfigMapList
except ImportError:  # TODO(python): packages under `io.` other than `io.micronaut` cannot be imported at runtime
    from kubernetes.client.informer import ResourceEventHandler
    from kubernetes.client.openapi.models import V1ConfigMap, V1ConfigMapList

LOG = logging.getLogger(__name__)


@Requires(property="spec.name", value="ConfigMapInformerSpec")
# tag::handler[]
@Informer(apiType=V1ConfigMap, apiListType=V1ConfigMapList)  # <1>
class ConfigMapInformer(ResourceEventHandler[V1ConfigMap]):  # <2>

    # end::handler[]
    def __init__(self):
        self.added: list[V1ConfigMap] = []
        self.updated: list[V1ConfigMap] = []
        self.deleted: list[V1ConfigMap] = []

    # tag::handler[]
    def onAdd(self, obj: V1ConfigMap) -> None:
        # end::handler[]
        LOG.info("ADDED CONFIG MAP: %s", obj)
        self.added.append(obj)
        # tag::handler[]
        ...

    def onUpdate(self, old_obj: V1ConfigMap, new_obj: V1ConfigMap) -> None:
        # end::handler[]
        LOG.info("UPDATED CONFIG MAP: %s", new_obj)
        self.updated.append(new_obj)
        # tag::handler[]
        ...

    def onDelete(self, obj: V1ConfigMap, deleted_final_state_unknown: bool) -> None:
        # end::handler[]
        LOG.info("DELETED CONFIG MAP: %s", obj)
        self.deleted.append(obj)
        # tag::handler[]
        ...
# end::handler[]
